proposals/044-integration-model.md

```diff
@@ -0,0 +1,203 @@
+* The _active_ integration's `targetState` is set to `Undeployed`
+* As soon this state has been reached (callback), this integration's `targetState` is set to `Active`.
+
+That way it is ensured that only a single version of an integration is active. 
+
+## Data model
+
+The following data model only highlights the parts which are specific to the lifecycle model:
+
+```java
+public class Integration {
+    
+    // Unique name of the integration
+    String name;
+    
+    // Current state of the integration
+    IntegrationState currentState;
+    
+    // Target state of the integration    
+    IntegrationState targetState;
+    
+    // The version of the integration which was the origin of this integration
+    // 0 if this is a new integration
+    int parentVersion;
+    
+    // Version of this integratin
+    int version;
+  
+    // Message describing the state further (e.g. error message)
+    String stateMessage;
+    
+    // Whether this integration has converged to its target state
+    boolean isPending() {
+        return currentState != targetState;
+    }
+}
+
+enum IntegrationState {
+  
+    // Initial state of an undeployed integration
+    Draft,
+  
+    // Deployed and running
+    Active, 
+  
+    // Deployed and suspended
+    Inactive,
+  
+    // Undeployed (no associated pods)
+    Undeployed,
+  
+    // Error occured
+    Error;
+}
+```
+
+## Deployment
+
+Deployment goes when an integration with `currentState == Draft` set `targetState = Active`. 
+Several actions will be performed by the controller:
+
+* A GitHub repository will be created if this is the first version of an integration (parentVersion == 0). 
+* If it is an new version of an existing integration (e.g. because a user changed something), a new Git branch is created from the old version's branch. This branch is named after the version.
+* The runtime files are created and commit to the Git repository
+* If this is an update of an existing integration, a currently running or suspended integration (`currentState == Active or Inactive`) is set to `targetState = Undeployed` (which triggers an undeployment on behalf of the controller)
+* The resource descriptors are created for this integration and applied to OpenShift.
+
+Open Question: For an update should an exisiting `DeploymentConfig` be reused (and e.g. doing a rolling update) or should we create a new `DeploymentConfig` ?
+
+## UI changes and updates
+
+Within the UI the state of an integration should be visualized. 
+It is suggested that when the reconciliation state is `Ok` the _currentState_ should be shown and when the reconciliation state is `Pending` then "Pending" should be shown (possibly whith the `targetState` as a tool tip info). Alternatively, when a reconciliation process the label could be more specific about the action just performing ("Deploying", "Suspending", "Resuming", "Undeploying") which can be infered from `currentState` and `targetState`
```

Comment:
Yeah, if we maybe have the reconciliation state as a separate field on the integration that'd make the logic in the UI easier.  Most likely when it's `Pending` then we'll show the `currentState` using the appropriate visualization.  I guess though the reconciliation state attribute would have a third state, `Error` so we can show whatever error information is available.